front/lib/actions/mcp_internal_actions/servers/trigger_management.ts

```diff
@@ -0,0 +1,539 @@
+import type { AgentLoopContextType } from "@app/lib/actions/types";
+import { generateCronRule } from "@app/lib/api/assistant/configuration/triggers";
+import type { Authenticator } from "@app/lib/auth";
+import { TriggerResource } from "@app/lib/resources/trigger_resource";
+import { rateLimiter } from "@app/lib/utils/rate_limiter";
+import { getStatsDClient } from "@app/lib/utils/statsd";
+import logger from "@app/logger/logger";
+import type { Result } from "@app/types";
+import { Err, Ok } from "@app/types";
+import type { AgentConfigurationType } from "@app/types/assistant/agent";
+import { isUserMessageType } from "@app/types/assistant/conversation";
+
+function renderTrigger(trigger: TriggerResource): string {
+  const config = trigger.configuration;
+  const scheduleInfo =
+    trigger.kind === "schedule" && "cron" in config
+      ? `${trigger.naturalLanguageDescription ?? config.cron} (${config.timezone})`
+      : trigger.kind;
+  const lines = [
+    `- **${trigger.name}** (ID: ${trigger.sId()})`,
+    `  Schedule: ${scheduleInfo}`,
+  ];
+  if (trigger.customPrompt) {
+    lines.push(`  Prompt: ${trigger.customPrompt}`);
+  }
+  lines.push(`  Enabled: ${trigger.enabled ? "Yes" : "No"}`);
+  return lines.join("\n");
+}
+
+function getToolContext(
+  auth: Authenticator,
+  agentLoopContext?: AgentLoopContextType
+): Result<
+  {
+    userId: number;
+    workspaceSId: string;
+    agentConfiguration: AgentConfigurationType;
+  },
+  MCPError
+> {
+  const user = auth.user();
+  const workspace = auth.workspace();
+  if (!user) {
+    return new Err(new MCPError("User not found"));
+  }
+  if (!workspace) {
+    return new Err(new MCPError("Workspace not found"));
+  }
+  if (!agentLoopContext?.runContext) {
+    return new Err(new MCPError("Agent context is required"));
+  }
+  return new Ok({
+    userId: user.id,
+    workspaceSId: workspace.sId,
+    agentConfiguration: agentLoopContext.runContext.agentConfiguration,
+  });
+}
+
+function getUserTimezone(
+  agentLoopContext?: AgentLoopContextType
+): string | null {
+  const content = agentLoopContext?.runContext?.conversation?.content;
+  if (!content) {
+    return null;
+  }
+
+  const userMessage = content.flat().findLast(isUserMessageType);
+  return userMessage?.context.timezone ?? null;
+}
+
+async function fetchTriggerWithOwnershipCheck(
+  auth: Authenticator,
+  triggerId: string,
+  agentConfigurationSId: string,
+  userId: number
+): Promise<Result<TriggerResource, MCPError>> {
+  const trigger = await TriggerResource.fetchById(auth, triggerId);
+  if (!trigger) {
+    return new Err(new MCPError("Trigger not found"));
+  }
+  if (trigger.agentConfigurationId !== agentConfigurationSId) {
+    return new Err(new MCPError("This trigger does not belong to this agent"));
+  }
+  if (trigger.editor !== userId) {
+    return new Err(new MCPError("You can only modify triggers you created"));
+  }
+  return new Ok(trigger);
+}
```

Comment:
ideally this lives in the resource right ? like a fetchByIdForEditor or smth similar